Jetpack: SSO security settings

tonyr59h opened the pull request:
This is the last of the security settings for Jetpack. Note that there will be at least two subsequent PR's to fix some issues with the branch before merging into `develop`. I'll outline the remaining work in the ticket (#6194).

This PR adds functionality to the final section of [this screen](https://cldup.com/Wu-h_dhFjj.png), with one major exception being that `Match accounts using email addresses` changes cannot be saved remotely. This is one of the future PR's 😉  I'm working with the API folks to figure out where the option is exposed.

To test:
* Login to Jetpack site, login on the web to help test
* Make sure the initial values are correct
* Toggle the settings (save the second one) and make sure they're updated remotely

cc @oguzkocer

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/prefs/DotComSiteSettings.java:
```diff
@@ -198,13 +197,147 @@ public void onErrorResponse(VolleyError error) {
+                    public void onResponse(JSONObject response) {
+                        AppLog.v(AppLog.T.API, "Received Jetpack SSO module");
+                        mRemoteJpSettings.ssoActive = response.optBoolean("active");
+                        mJpSettings.ssoActive = response.optBoolean("active");
```
We should be consistent with other fetch methods and use:
```
mJpSettings.ssoActive = mRemoteJpSettings.ssoActive;
```

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/prefs/DotComSiteSettings.java:
```diff
@@ -198,13 +197,147 @@ public void onErrorResponse(VolleyError error) {
+                    public void onResponse(JSONObject response) {
+                        AppLog.v(AppLog.T.API, "Received Jetpack Protect module");
+                        mRemoteJpSettings.jetpackProtectEnabled = response.optBoolean("active");
+                        mJpSettings.jetpackProtectEnabled = mRemoteJpSettings.jetpackProtectEnabled;
```
Should we have `notifyUpdatedOnUiThread(null);` after this?

tonyr59h replied:
Added in 6bf2d0c0aa858fad9987f6927647dfe291193f01

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/prefs/DotComSiteSettings.java:
```diff
@@ -198,13 +197,147 @@ public void onErrorResponse(VolleyError error) {
+                                    mJpSettings.ssoMatchEmail = mRemoteJpSettings.ssoMatchEmail;
+                                }
+                            } catch (JSONException e) {
+                                e.printStackTrace();
```
We probably should at least log the exception.

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/prefs/DotComSiteSettings.java:
```diff
@@ -198,13 +197,147 @@ public void onErrorResponse(VolleyError error) {
+                        AppLog.v(AppLog.T.API, "Received Jetpack SSO module 2FA option");
+                        mRemoteJpSettings.ssoRequireTwoFactor = response.optBoolean("option_value");
+                        mJpSettings.ssoRequireTwoFactor = mRemoteJpSettings.ssoRequireTwoFactor;
+                            notifyUpdatedOnUiThread(null);
```
There is an extra tab space here.

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/prefs/DotComSiteSettings.java:
```diff
@@ -216,12 +349,125 @@ public void onResponse(JSONObject response) {
+                    @Override
+                    public void onErrorResponse(VolleyError error) {
+                        AppLog.w(AppLog.T.API, "Error updating Jetpack Monitor module: " + error);
+                        notifySavedOnUiThread(error);
```
This is a bit inconsistent between calls. Some of the errors have this call and most don't. Unless there is a reason, we should be consistent everywhere. If there is a reason, I think it's worth a comment when it's required.

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/prefs/DotComSiteSettings.java:
```diff
@@ -216,12 +349,125 @@ public void onResponse(JSONObject response) {
+        final Map<String, String> params = serializeJetpackParams();
+        if (params == null || params.isEmpty()) return;
+
+        WordPress.getRestClientUtils().setJetpackSettings(
```
I am not quite sure why this is here inside a method called `pushJetpackMonitorSettings`.

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/prefs/DotComSiteSettings.java:
```diff
@@ -216,12 +349,125 @@ public void onResponse(JSONObject response) {
+                        AppLog.w(AppLog.T.API, "Error updating Jetpack Monitor module options: " + error);
+                        notifySavedOnUiThread(error);
+                    }
+                }, params);
```
I thought we fixed this in another PR, but maybe it was a different method. We should always have the callbacks as the last parameters for ease of read. Can we move the params to after site id parameter?

tonyr59h commented:
Thanks for the review!

> All these calls to different API endpoints to update single settings is .. something.

Yeah, I've had a rough week tracking down which endpoints return which settings. After chatting with JPOP it seems there may be a better way to do all of this. I'm going to try and get their suggestions working since it would greatly simplify the network calls. This PR should still get merged. I'll open another if I manage to get everything working.

> One thing I noticed is that we don't have any error handling. Maybe it was intentional, since we have so many calls that might be happening at the same time, so if we did, we might end up showing a bunch of errors to the user.

While testing I'd run into cases where 5+ toasts would queue up. I added a flag to make sure only one error shows a toast but I didn't want to bloat this PR anymore. I've updated #6440 to include error handling, let's do it in a separate PR.

tonyr59h commented:
Alright @oguzkocer, I've made some big changes to update the endpoints we're using. We're down to just two!

`v1.1/jetpack-blogs/$site_id` is used to fetch/push _only_ Jetpack Monitor settings, but not to actually toggle Jetpack Monitor. For some reason pushing `monitor_active` is successful but it doesn't actually disable the module.

`v1.1/jetpack-blogs/$site_id/rest-api/?path=/jetpack/v4/settings/` is used to fetch/push all module (de)activations, Protect settings, and SSO settings.

The latter endpoint is probably going to be very useful for us as it allows us to perform authenticated WP REST queries on Jetpack sites using WP.com credentials.

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/prefs/DotComSiteSettings.java:
```diff
@@ -198,13 +159,87 @@ public void onErrorResponse(VolleyError error) {
+                        for (int i = 0; i < whitelistItems.length(); ++i) {
+                            String item = whitelistItems.optString(i, "");
+                            if (!item.isEmpty() && !mRemoteJpSettings.jetpackProtectWhitelist.contains(item)) {
+                                mRemoteJpSettings.jetpackProtectWhitelist.add(item);
```
Shouldn't we clear the `mRemoteJpSettings.jetpackProtectWhitelist` before all this?

tonyr59h replied:
Done in b1cf0912b118a3a96749a0a88b502a5701de8701

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/prefs/DotComSiteSettings.java:
```diff
@@ -216,12 +251,106 @@ public void onResponse(JSONObject response) {
+    private void pushJetpackProtectAndSsoSettings() {
+        final Map<String, Object> params = serializeJetpackProtectAndSsoParams();
+        if (params.isEmpty()) {
+            return;
```
We might want to add a log here for HEs.

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/prefs/DotComSiteSettings.java:
```diff
@@ -185,8 +146,8 @@ public void onResponse(JSONObject response) {
                         }
 
                         if (mSite.isJetpackConnected()) {
```
Is there a reason these calls are not being made in `fetchRemoteData`?

tonyr59h replied:
I added it in here originally to prevent multiple error toasts when fetching. I'll move this block to `fetchRemoteData` and figure out a better way to handle the multiple error toasts in a future PR.

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/prefs/DotComSiteSettings.java:
```diff
@@ -216,12 +251,106 @@ public void onResponse(JSONObject response) {
+                    @Override
+                    public void onResponse(JSONObject response) {
+                        AppLog.d(AppLog.T.API, "Jetpack Monitor module updated");
+                        mRemoteJpSettings.emailNotifications = mJpSettings.emailNotifications;
```
This doesn't look correct. Why are we not using the response object here?

tonyr59h replied:
This response also contains no useful information. I'll add some comments to these two `onResponse` methods to clarify why the response object is not used.

![](https://cldup.com/bX8q6I0pRR.png)

oguzkocer replied:
Same as [this](https://github.com/wordpress-mobile/WordPress-Android/pull/6440#discussion_r134191075)

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/prefs/DotComSiteSettings.java:
```diff
@@ -404,169 +533,67 @@ private JSONObject serializeDotComParamsToJSONObject() throws JSONException {
         Map<String, String> params = new HashMap<>();
-        params.put(JP_MONITOR_EMAIL_NOTES_KEY, String.valueOf(mJpSettings.emailNotifications));
-        params.put(JP_MONITOR_WP_NOTES_KEY, String.valueOf(mJpSettings.wpNotifications));
+        if (mJpSettings.emailNotifications != mRemoteJpSettings.emailNotifications) {
```
I personally would prefer not to make these checks because there is no guarantee that the remote settings is not changed since our last fetch/push and we don't really lose anything by updating these settings even if they are the same. That goes for everything in `serializeJetpackProtectAndSsoParams` too. I do see the logic behind it, so I am not saying we should change it, but just mentioning it as something to consider.

tonyr59h replied:
We're using these checks to prevent unnecessary network calls (see `pushJetpackProtectAndSsoSettings` and `pushWpSettings`). I think we can go about preventing those calls just using `equals` which would allow us to remove these checks, I'd prefer to tackle that in a separate PR.

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/prefs/DotComSiteSettings.java:
```diff
@@ -216,12 +251,106 @@ public void onResponse(JSONObject response) {
+                    @Override
+                    public void onResponse(JSONObject response) {
+                        AppLog.d(AppLog.T.API, "Jetpack settings updated");
+                        mRemoteJpSettings.monitorActive = mJpSettings.monitorActive;
```
This doesn't look correct. Why are we not using the response object here?

tonyr59h replied:
The response object doesn't actually contain anything of interest :/

![](https://cldup.com/g13aVoL3sT.png)

oguzkocer replied:
In that case, I think we should just copy the values in `mJpSettings` just before the call and use those values in the callback as there is no guarantee that `mJpSettings` is not modified during the call.

tonyr59h replied:
Good idea 👍

tonyr59h commented:
> Inconsistency in the usage of notifyUpdatedOnUiThread(null);. Some remote calls have it, some don't. Some error callbacks have it and some don't. If they are correct, it's not very clear. I'd consider moving that single call to a method and add some comments there about when it should or shouldn't be called.

Yeah, I tinkered with these `notify*` method calls due to multiple toasts being shown but I never went back to make their usage consistent. `notifyUpdatedOnUiThread(null)` needs to be called to get the UI updated, `notifyUpdatedOnUiThread(error)` to present a toast. I made some changes in 0f3454ba5c43c0211b734642bc54c8f7bb2adb98 to consistently call these in all `fetch*` and `push*` methods. I think this will lead to multiple toasts, but improved error handling is on the TODO list so I'll fix that in a future PR.

> I believe fetch and push callbacks should be handled the same way. The response should always supersede the local variables.

I think this is a reference to [this comment](https://github.com/wordpress-mobile/WordPress-Android/pull/6440#discussion_r131385004). I agree with the bullet point, but it's not possible for some of the responses we receive 😞 

> We assume callbacks happen immediately in some cases which can become a problem.

I tried to remedy this by only modifying fields in the callbacks that would be affected by the specific network call. Did you run into a case where the settings values weren't what you expected?

oguzkocer commented on libs/networking/WordPressNetworking/src/main/java/org/wordpress/android/networking/RestClientUtils.java:
```diff
@@ -196,47 +198,45 @@ public void getCurrentTheme(long siteId, Listener listener, ErrorListener errorL
+            params.put("body", body.toString());
+            post(path, params, null, listener, errorListener);
+        } catch (JSONException e) {
+            AppLog.e(AppLog.T.API, "Error updating Jetpack settings: " + e);
```
I think I missed this in my previous review. We should call the error listener here.

tonyr59h commented:
> We are almost ready to merge this in. I have left a comment about an error callback and followed up on the response comment here, once these 2 are resolved we can merge it in.

All patched up! Thanks for the detailed review, now to cleanup the branch 😄
